lib/utils/fncache_test.go

```diff
@@ -56,10 +60,99 @@ func TestFnCache_New(t *testing.T) {
 	}
 }
 
-// TestFnCacheSanity runs basic FnCache test cases.
-func TestFnCacheSanity(t *testing.T) {
+type result struct {
+	val interface{}
+	err error
+}
+
+// TestFnCacheConcurrentReads verifies that many concurrent reads result in exactly one
+// value being actually loaded via loadfn if a reasonably long TTL is used.
+func TestFnCacheConcurrentReads(t *testing.T) {
+	const workers = 100
+	t.Parallel()
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// set up a chage that won't ttl out values during the test
+	cache, err := NewFnCache(FnCacheConfig{TTL: time.Hour})
+	require.NoError(t, err)
+
+	results := make(chan result, workers)
+
+	for i := 0; i < workers; i++ {
+		go func(n int) {
+			val, err := cache.Get(ctx, "key", func(context.Context) (interface{}, error) {
+				// return a unique value for each worker so that we can verify whether
+				// the values we get come from the same loadfn or not.
+				return fmt.Sprintf("val-%d", n), nil
+			})
+			results <- result{val, err}
+		}(i)
+	}
+
+	first := <-results
+	require.NoError(t, first.err)
+
+	val := first.val.(string)
+	require.NotZero(t, val)
+
+	for i := 0; i < (workers - 1); i++ {
+		r := <-results
+		require.NoError(t, r.err)
+		require.Equal(t, val, r.val.(string))
+	}
+}
+
+// TestFnCacheExpiry verfies basic expiry.
+func TestFnCacheExpiry(t *testing.T) {
 	t.Parallel()
 
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
```

Comment:
nit:
```suggestion
	t.Cleanup(cancel)
```

Reply:
IMO `defer` is still the preferable way to do things like this.  `t.Cleanup` is great for simplifying the API of testing helpers/suites, and for handling the cleanup of resources shared across subtests, but IMO replacing a normal `defer` with a `t.Cleanup` is pointless and potentially harmful since it creates an unnecessary divergence between test and non-test code styles.